File: bilibili/go_basic_zcy/func_var_demo.go
package main

import "fmt"

/*
对比go和python函数传参的底层逻辑
https://www.bilibili.com/video/BV1u33TzEE83/
*/

func foo(n int) {
	fmt.Printf("n=%d, n的地址：%p\n", n, &n)
	n++
	fmt.Printf("n=%d, n的地址：%p\n", n, &n)
}

func main() {
	i := 8
	fmt.Printf("i=%d, i的地址：%p\n", i, &i)
	foo(i) // go语言这里实际上是把备份传给了foo函数，所以函数内的变量和函数外的变量地址不同。
	fmt.Printf("i=%d, i的地址：%p\n", i, &i)
}

/*
i=8, i的地址：0xc00000a0f8
n=8, n的地址：0xc00000a118
n=9, n的地址：0xc00000a118
i=8, i的地址：0xc00000a0f8
*/

File: bilibili/go_basic_zcy/func_var_demo/func_var_demo.py
def foo(n: int) -> None:
    print(f"n={n}, n的地址 {id(n)}")
    n += 1   # 因为python中int类型是常量，所以n+=1需要copy一份出来
    print(f"n={n}, n的地址 {id(n)}")  # 所以这里的打印id不一样了

if __name__ == '__main__':
    i = 8
    print(f"i={i}, i的地址 {id(i)}")
    foo(i)
    print(f"i={i}, i的地址 {id(i)}")

"""
i=8, i的地址 140714461186712
n=8, n的地址 140714461186712
n=9, n的地址 140714461186744
i=8, i的地址 140714461186712
"""